lib/datadog/tracing/contrib/aws/service/states.rb

```diff
@@ -16,21 +16,35 @@ def add_tags(span, params)
               execution_arn = params[:execution_arn]
               state_machine_account_id = ''
 
+              if state_machine_arn
+                span.set_tag(Aws::Ext::TAG_STATE_MACHINE_ARN, state_machine_arn)
+                # https://docs.aws.amazon.com/step-functions/latest/apireference/API_StartExecution.html#API_StartExecution_RequestSyntax:~:text=Required%3A%20No-,stateMachineArn,-The%20Amazon%20Resource
+                # arn:<partition>:states:<region>:<account-id>:stateMachine:<myStateMachineName>
+                # arn:<partition>:states:<region>:<account-id>:stateMachine:<myStateMachineName>:10
+                # arn:<partition>:states:<region>:<account-id>:stateMachine:<myStateMachineName:PROD>
+                # There are 3 patterns to cover and attempt to capture the `myStateMachineName`, it should always be in index 6 and account_id at index 4
+                parts = state_machine_arn.split(':')
+                state_machine_name ||= parts[6]
+                state_machine_account_id = parts[4]
+              end
+
               if execution_arn
-                # 'arn:aws:states:us-east-1:123456789012:execution:example-state-machine:example-execution'
+                span.set_tag(Aws::Ext::TAG_STATE_EXECUTION_ARN, execution_arn)
+                # express
+                # arn:aws:states:sa-east-1:123456789012:express:targetStateMachineName:1234:5678
+                # standard
+                # arn:aws:states:sa-east-1:123456789012:execution:targetStateMachineName:1234
                 parts = execution_arn.split(':')
-                state_machine_name = parts[-2]
+                state_machine_name = parts[6]
                 state_machine_account_id = parts[4]
               end
-
-              if state_machine_arn
-                # example statemachinearn: arn:aws:states:us-east-1:123456789012:stateMachine:MyStateMachine
-                parts = state_machine_arn.split(':')
-                state_machine_name ||= parts[-1]
-                state_machine_account_id = parts[-3]
+              
+              if state_machine_account_id
```

Comment:
`if state_machine_account_id` will always be true because `state_machine_account_id` was initialized with a non-nil value `state_machine_account_id = ''`.

Also, we don't want to set an empty value to a span tag (if `state_machine_account_id == ''`).

I recommend not initializing `state_machine_account_id = ''` at the beginning of this method (by default, it will be `state_machine_account_id == nil` on first access if never written to).

Reply:
nice catch, updated to explicitly define as nil